File: NetFramework/Pdf/VB/CoolPrintPreview/PrintDocumentPdfExporter.vb
Imports System.Drawing
Imports System.Drawing.Printing
Imports System.Collections.Generic
Imports System.Windows.Forms
Imports System.Text

Imports C1.C1Pdf

''' <summary>
''' Represents a class that exports a <see cref="PrintDocument"/> to a
''' PDF file.
''' </summary>
Friend Class PrintDocumentPdfExporter
	'-----------------------------------------------------------
	#Region "** fields"

	Private _doc As PrintDocument
	Private _pdf As C1PdfDocument
	Private _previewController As PreviewPrintController
	Private _pageCount As Integer
	Private _cancel As Boolean

	#End Region

	'-----------------------------------------------------------
	#Region "** ctor"

	Public Sub New(pdf As C1PdfDocument)
		' save reference to pdf component
		_pdf = pdf
	End Sub

	#End Region

	'-----------------------------------------------------------
	#Region "** object model"

	Public Function RenderDocument(doc As PrintDocument, showProgressDialog As Boolean) As Boolean
		' save reference to document
		_doc = doc

		' initialize pdf document
		_pdf.Clear()
		_pdf.Landscape = False

		' prepare to render
        Dim savePC As PrintController = _doc.PrintController
		_previewController = New PreviewPrintController()
		_doc.PrintController = If(showProgressDialog, New PrintControllerWithStatusDialog(_previewController), DirectCast(_previewController, PrintController))
		_pageCount = 0
		_cancel = False

		' render
		Try
			AddHandler _doc.PrintPage, AddressOf _doc_PrintPage
			AddHandler _doc.EndPrint, AddressOf _doc_EndPrint
			_doc.Print()
		Finally
			RemoveHandler _doc.PrintPage, AddressOf _doc_PrintPage
			RemoveHandler _doc.EndPrint, AddressOf _doc_EndPrint
			_doc.PrintController = savePC
		End Try

		' done
		Return Not _cancel
	End Function

	#End Region

	'-----------------------------------------------------------
	#Region "** PrintDocument event handlers"

	Private Sub _doc_PrintPage(sender As Object, e As PrintPageEventArgs)
        Dim pages As PreviewPageInfo() = _previewController.GetPreviewPageInfo()
        While _pageCount < pages.Length - 1
            DrawPage(pages, System.Math.Max(_pageCount, _pageCount - 1))
            _pageCount = _pageCount + 1
        End While
        If e.Cancel Then
            _cancel = True
        End If
    End Sub
	Private Sub _doc_EndPrint(sender As Object, e As PrintEventArgs)
        Dim pages As PreviewPageInfo() = _previewController.GetPreviewPageInfo()
		DrawPage(pages, _pageCount)
		If e.Cancel Then
			_cancel = True
		End If
	End Sub
    Private Sub DrawPage(pages As PreviewPageInfo(), index As Integer)

        ' skip to next page
        If index > 0 Then
            _pdf.NewPage()
        End If

        ' get preview page info
        Dim pi As PreviewPageInfo = pages(index)

        ' adjust page size
        Dim ps As Size = pi.PhysicalSize
        _pdf.PageSize = New SizeF(ps.Width / 100.0F * 72, ps.Height / 100.0F * 72)

        ' draw image
        Dim img As Image = pi.Image
        _pdf.DrawImage(img, _pdf.PageRectangle)
    End Sub

	#End Region
End Class

File: NetFramework/FlexGrid/CS/MultiGridDocument/CoolPrintPreview/PrintDocumentPdfExporter.cs
using System;
using System.Drawing;
using System.Drawing.Printing;
using System.Collections.Generic;
using System.Windows.Forms;
using System.Text;

namespace CoolPrintPreview
{
    using C1.C1Pdf;

    /// <summary>
    /// Represents a class that exports a <see cref="PrintDocument"/> to a
    /// PDF file.
    /// </summary>
    internal class PrintDocumentPdfExporter
    {
        //-----------------------------------------------------------
        #region ** fields

        PrintDocument _doc;
        C1PdfDocument _pdf;
        PreviewPrintController _previewController;
        int _pageCount;
        bool _cancel;

        #endregion

        //-----------------------------------------------------------
        #region ** ctor

        public PrintDocumentPdfExporter(C1PdfDocument pdf)
        {
            // save reference to pdf component
            _pdf = pdf;
        }

        #endregion

        //-----------------------------------------------------------
        #region ** object model

        public bool RenderDocument(PrintDocument doc, bool showProgressDialog)
        {
            // save reference to document
            _doc = doc;

            // initialize pdf document
            _pdf.Clear();
            _pdf.Landscape = false;

            // prepare to render
            var savePC = _doc.PrintController;
            _previewController = new PreviewPrintController();
            _doc.PrintController = showProgressDialog
                ? new PrintControllerWithStatusDialog(_previewController)
                : (PrintController)_previewController;
            _pageCount = 0;
            _cancel = false;

            // render
            try
            {
                _doc.PrintPage += _doc_PrintPage;
                _doc.EndPrint += _doc_EndPrint;
                _doc.Print();
            }
            finally
            {
                _doc.PrintPage -= _doc_PrintPage;
                _doc.EndPrint -= _doc_EndPrint;
                _doc.PrintController = savePC;
            }

            // done
            return !_cancel;
        }

        #endregion

        //-----------------------------------------------------------
        #region ** PrintDocument event handlers

        void _doc_PrintPage(object sender, PrintPageEventArgs e)
        {
            var pages = _previewController.GetPreviewPageInfo();
            while (_pageCount < pages.Length - 1)
            {
                DrawPage(pages, _pageCount++);
            }
            if (e.Cancel)
            {
                _cancel = true;
            }
        }
        void _doc_EndPrint(object sender, PrintEventArgs e)
        {
            var pages = _previewController.GetPreviewPageInfo();
            DrawPage(pages, _pageCount);
            if (e.Cancel)
            {
                _cancel = true;
            }
        }
        void DrawPage(PreviewPageInfo[] pages, int index)
        {
            // skip to next page
            if (index > 0)
            {
                _pdf.NewPage();
            }

            // get preview page info
            var pi = pages[index];

            // adjust page size
            var ps = pi.PhysicalSize;
            _pdf.PageSize = new SizeF(ps.Width / 100f * 72, ps.Height / 100f * 72);

            // draw image
            var img = pi.Image;
            _pdf.DrawImage(img, _pdf.PageRectangle);
        }

        #endregion
    }
}
